pallets/orml-tokens-extension/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+		AlreadyCreated,
+		/// Trying to register a currency variant that is not ownable
+		NotOwnableCurrency,
+		/// Currency has not been created
+		NotCreated,
+		/// No permission to call the operation
+		NoPermission,
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	// The pallet's dispatchable functions.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create and take ownership of one CurrencyId
+		///
+		/// The creator will have full control of this pallet's functions
+		/// regarding this currency
+		///
+		/// Parameters:
+		/// - `currency_id`: Allowed Currency Id.
+		///
+		/// Emits `Created` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::create())]
+		#[transactional]
+		pub fn create(origin: OriginFor<T>, currency_id: CurrencyOf<T>) -> DispatchResult {
+			let creator = ensure_signed(origin)?;
+
+			ensure!(
+				T::CurrencyIdChecker::is_valid_currency_id(&currency_id),
+				Error::<T>::NotOwnableCurrency
+			);
+			ensure!(!CurrencyData::<T>::contains_key(&currency_id), Error::<T>::AlreadyCreated);
+
+			CurrencyData::<T>::insert(
+				currency_id.clone(),
+				CurrencyDetails {
+					owner: creator.clone(),
+					issuer: creator.clone(),
+					admin: creator.clone(),
+				},
+			);
+
+			Self::deposit_event(Event::Created {
+				currency_id,
+				creator: creator.clone(),
+				owner: creator,
+			});
+
+			Ok(())
+		}
+
+		/// Mint currency of a particular class.
+		///
+		/// The origin must be Signed and the sender must be the Issuer of the currency `id`.
+		///
+		/// - `id`: The identifier of the currency to have some amount minted.
+		/// - `beneficiary`: The account to be credited with the minted currency.
+		/// - `amount`: The amount of the currency to be minted.
+		///
+		/// Emits `Issued` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as Config>::WeightInfo::mint())]
+		#[transactional]
+		pub fn mint(
+			origin: OriginFor<T>,
+			currency_id: CurrencyOf<T>,
+			to: AccountIdOf<T>,
+			amount: BalanceOf<T>,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			// get currency details and check issuer
+			let currency_data =
+				CurrencyData::<T>::get(currency_id).ok_or(Error::<T>::NotCreated)?;
+			ensure!(origin == currency_data.issuer, Error::<T>::NoPermission);
+
+			// do mint via orml-currencies
+			let _ = ext::orml_tokens::mint::<T>(currency_id, &to, amount)?;
+
+			Self::deposit_event(Event::Mint { currency_id, to, amount });
+			Ok(())
+		}
+
+		/// Burn currency of a particular class.
+		///
+		/// The origin must be Signed and the sender must be the Admin of the currency `id`.
+		///
+		/// - `id`: The identifier of the currency to have some amount burned.
+		/// - `from`: The account to be debited.
+		/// - `amount`: The amount of the currency to be burned.
+		///
+		/// Emits `Burned` event when successful.
+		///
+		/// Weight: `O(1)`
```

Comment:
Maybe we should remove all these 'Weight' comments now that we have proper benchmarks